src/plugins/kibana/server/routes/api/ingest/register_bulk.js

```diff
@@ -0,0 +1,98 @@
+import { Promise } from 'bluebird';
+import { parse, transform } from 'csv';
+import _ from 'lodash';
+import hi from 'highland';
+import { patternToIngest } from '../../../../common/lib/convert_pattern_and_ingest_name';
+import { PassThrough } from 'stream';
+import bulkRequestSchema from '../../../lib/schemas/bulk_request_schema';
+
+export function registerBulk(server) {
+  server.route({
+    path: '/api/kibana/{id}/_bulk',
+    method: 'POST',
+    config: {
+      payload: {
+        output: 'stream',
+        maxBytes: 1024 * 1024 * 1024
+      },
+      validate: {
+        payload: bulkRequestSchema
+      }
+    },
+    handler: function (req, reply) {
+      const boundCallWithRequest = _.partial(server.plugins.elasticsearch.callWithRequest, req);
+      const indexPattern = req.params.id;
+      const usePipeline = req.payload.pipeline;
+      const csv = req.payload.csv;
+      const fileName = csv.hapi.filename;
+      const responseStream = new PassThrough();
+      const parser = parse({
+        columns: true,
+        auto_parse: true,
+        delimiter: _.get(req.payload, 'delimiter', ',')
+      });
+
+      let currentLine = 2; // Starts at 2 since we parse the header separately
+
+      responseStream.write('[');
+      csv.pipe(parser);
+
+      hi(parser)
```

Comment:
Yeah, to be clear: the standalone library thing makes implementations easier to rationalize for me, but the term "standalone library" seems to be the thing most at contention here, so let me put my feedback in a more useful way: I think this code would be easier to understand and less prone to regressions if it were pulled out into a JS module with clear boundaries and with unit test coverage.  It wouldn't need to be contributed back to elasticsearch-js or be published as its own library to address those two things.

My personal preference is to publish that type of module as a standalone library so it can be versioned and whatnot, but that really isn't my goal at this point.

To address some of the other points:

> I wrote some pretty extensive functional tests that cover everything the unit tests would, and more

Functional tests and unit tests don't test for the same thing, so having one doesn't really mean the other is less important.  Functional tests are great at showing me that this feature as a whole works as a user would expect at this moment, but they don't increase my confidence that changes to this code in the future aren't introducing unexpected side effects, but unit tests would.

> I was actually thinking the "streaming bulk" implementation might belong in the official elasticsearch js client.

I don't mean to pull this back to the standalone library thing, but I just wanted to say that I don't disagree with you there, though I do have some ideas about that.  I'd be happy to chat about it if we were ever considering the standalone library route.